spyder/api/plugins.py

```diff
@@ -16,232 +16,184 @@
 import os
 
 # Third party imports
-from qtpy.QtCore import Qt, Signal
-from qtpy.QtGui import QCursor
-from qtpy.QtWidgets import (QApplication, QMenu, QMessageBox, QToolButton,
-                            QWidget)
+from qtpy.QtCore import Signal, Slot
+from qtpy.QtWidgets import QWidget
 
 # Local imports
-from spyder.config.base import _
-from spyder.config.gui import get_color_scheme, is_dark_interface
-from spyder.config.main import CONF
 from spyder.config.user import NoDefault
-from spyder.plugins.base import BasePluginMixin
-from spyder.py3compat import configparser
+from spyder.plugins.base import BasePluginMixin, BasePluginWidgetMixin
 from spyder.utils import icon_manager as ima
-from spyder.utils.qthelpers import (add_actions, create_toolbutton,
-                                    MENU_SEPARATOR, toggle_actions)
 
 
-class PluginWidget(QWidget, BasePluginMixin):
+# =============================================================================
+# SpyderPlugin
+# =============================================================================
+class BasePlugin(BasePluginMixin):
     """
-    Public interface for Spyder plugins.
+    Basic functionality for Spyder plugins.
 
-    Warning: Don't override any methods present here!
-
-    Signals:
-      * sig_option_changed
-          Example:
-            plugin.sig_option_changed.emit('show_all', checked)
-      * sig_show_message
-      * sig_update_plugin_title
+    WARNING: Don't override any methods or attributes present here!
     """
 
-    sig_option_changed = Signal(str, object)
+    # Use this signal to display a message in the status bar
     sig_show_message = Signal(str, int)
-    sig_update_plugin_title = Signal()
 
-    def __init__(self, main=None):
-        """Bind widget to a QMainWindow instance."""
-        super(PluginWidget, self).__init__(main)
-        assert self.CONF_SECTION is not None
+    # Use this signal to inform another plugin that a configuration
+    # value has changed.
+    sig_option_changed = Signal(str, object)
 
-        self.dockwidget = None
-        self.undocked_window = None
+    def __init__(self, parent=None):
+        super(BasePlugin, self).__init__(parent)
 
-        # Check compatibility
-        check_compatibility, message = self.check_compatibility()
-        if not check_compatibility:
-            self.show_compatibility_message(message)
+        # This is the plugin parent, which corresponds to the main
+        # window.
+        self.main = parent
 
+        # Filesystem path to this plugin
         self.PLUGIN_PATH = os.path.dirname(inspect.getfile(self.__class__))
```

Comment:
Yes, it's necessary to expose it in case plugins need to load some assets placed in directories relative to `PLUGIN_PATH`.

However, I moved the computation of `PLUGIN_PATH` to a method of `BasePluginMixin` called `_get_plugin_path` to not expose how we compute the method directly.